src/macaron/build_spec_generator/reproducible_central/reproducible_central.py

```diff
@@ -0,0 +1,471 @@
+            "Cannot find an analysis result for PackageURL %s in the database. "
+            + "Please check if an analysis for it exists in the database.",
+            purl.to_string(),
+        )
+        return None
+
+    latest_component_repository = latest_component.repository
+    if not latest_component_repository:
+        logger.error(
+            "Cannot find any repository information for %s in the database.",
+            purl.to_string(),
+        )
+        return None
+    logger.info(
+        "Repository information for purl %s: url %s, commit %s",
+        purl,
+        latest_component_repository.remote_path,
+        latest_component_repository.commit_sha,
+    )
+
+    # Getting the RC build tool name from the build tool check facts.
+    rc_build_tool_name = get_rc_build_tool_name(
+        component_id=latest_component.id,
+        session=session,
+    )
+    if not rc_build_tool_name:
+        return None
+
+    # We always attempt to get the JDK version from maven central JAR for this GAV artifact.
+    jdk_from_jar = find_jdk_version_from_central_maven_repo(
+        group_id=group,
+        artifact_id=artifact,
+        version=version,
+    )
+    logger.info(
+        "Attempted to find JDK from Maven Central JAR. Result: %s",
+        jdk_from_jar or "Cannot find any.",
+    )
+
+    # Obtain the highest confidence build command info from the database.
+    lookup_build_command_info = get_lookup_build_command_info(
+        component_id=latest_component.id,
+        session=session,
+    )
+    logger.info(
+        "Attempted to find build command from the database. Result: %s",
+        lookup_build_command_info or "Cannot find any.",
+    )
+    lookup_build_command_jdk = (
+        get_lookup_build_command_jdk(
+            lookup_build_command_info,
+        )
+        if lookup_build_command_info
+        else None
+    )
+
+    # Select jdk from jar from different source.
+    # The default JDK version is 8.
+    selected_jdk_version = jdk_from_jar or lookup_build_command_jdk or "8"
+    major_jdk_version = normalize_jdk_version(selected_jdk_version)
+    if not major_jdk_version:
+        logger.error("Failed to obtain the major version of %s", selected_jdk_version)
+        return None
+
+    # Select build commands from lookup or use a default one.
+    selected_build_command = (
+        lookup_build_command_info.command
+        if lookup_build_command_info
+        else get_rc_default_build_command(
+            rc_build_tool_name,
+        )
+    )
+    if not selected_build_command:
+        logger.error("Failed to get a build command for %s.", purl.to_string())
+        return None
+
+    patched_build_commands = patch_commands(
+        cmds_sequence=[selected_build_command],
+        patches=patches,
+    )
+    if not patched_build_commands:
+        logger.error(
+            "Failed to patch command sequences %s.",
+            [selected_build_command],
+        )
+        return None
```

Comment:
As the JDK discovery can rely on the build information via `lookup_build_command_info`, I'd suggest moving all of the build command related code before the JDK related code. Up to and including the highlighted block here.
Also might be a good idea to extract some of this out to standalone functions. E.g. JDK, Patched Build Commands, etc.